### ISSUE TITLE
arXiv '23 | Adding Conditional Control to Text-to-Image Diffusion Models.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#72]
- Zhang L, Agrawala M. [Adding Conditional Control to Text-to-Image Diffusion Models](https://arxiv.org/abs/2302.05543).
 - Zhang L. [Ablation Study: Why ControlNets use deep encoder? What if it was lighter? Or even an MLP?](https://github.com/lllyasviel/ControlNet/discussions/188)

**Comment by NorbertZheng** [comment]
Ablation Study: Why ControlNets use deep encoder? What if it was lighter? Or even an MLP?

**Comment by NorbertZheng** [comment]
## Overview
In 2023, if we want to train an encoder to perform some tasks, we have four basic options as follows:
 1. Train a lightweight encoder from scratch.
 2. Train a lightweight encoder by fine-tuning existing encoders.
 3. Train a deep encoder from scratch.
 4. Train a deep encoder by fine-tuning existing encoders.

In our problem, we want to control Stable Diffusion, and the encoder will be trained jointly with a big Stable Diffusion (SD). Because of this, ~~the option (3) requires super large computation power and is not practical unless you have [as many A100s as EMostaque does](https://twitter.com/EMostaque/status/1628235860670906368)~~. But we do not have that, so we may just forget about (3).

The option (1) and (2) are just similar and can be merged. They usually have similar performances.

Note that
 - fine-tuning existing deep encoders,
 - and training lightweight encoders from scratch

are both relatively preferred methods. Which one is "harder" or "easier" to train is a complicated question and even related to different training environments. We should not ~~presume the learning behaviors by simply looking at the number of parameters~~.

But in this post, let's pay more attention to the qualitative differences of these methods if they are already trained successfully.

**Comment by NorbertZheng** [comment]
## Candidates
Let us consider these architectures:
### ControlNet-Self (Finetuned)
Below is the model architecture that we released many days ago as
 - a final solution.

It directly uses the encoder of Stable Diffusion (SD). Because it copies itself, let us call it **ControlNet-Self**.

![221408908-57480f60-55ca-404c-83ec-d1ac03c7f7f1](https://user-images.githubusercontent.com/37235023/222045593-51511f43-847b-4374-b5e5-24dae086e6db.png)

### ControlNet-Lite (From Scratch)
Below is a typical architecture to **train lightweight encoders from scratch**. We just use some simple convolution layers to
 - get some embedding,
 - and inject the Stable Diffusion (SD) U-Net.

Because it has relatively fewer parameters, let's call it **ControlNet-Lite**. Channels of layers are computed by instantiating the `ldm` python object.

![](https://user-images.githubusercontent.com/19834515/221408921-068dd793-53cf-4020-b542-403de295c04f.png)

### ControlNet-MLP (From Scratch)
Below is a more extreme case to just use
 - a pixel-wise Multilayer Perceptron (MLP).

In recent years, MLPs are suddenly popular again, and they are actually just $1\times 1$ convolutions. We use AVG pool as downsampling and then let us call it "ControlNet-MLP". Channels of layers are computed by instantiating the `ldm` python object.

![221408927-300c4f15-4c39-4fb6-9864-1b819e8c28d8](https://user-images.githubusercontent.com/37235023/222046215-731063cc-520a-4a94-8105-46d5248a3a98.png)

**Comment by NorbertZheng** [comment]
## Here We Go!
[This house image](https://www.pinterest.com/pin/8655424277388928/) is just the first searching result when I search "house" in pinterest. Let us use it as an example:
![221408945-7fe130a8-2a9b-4019-ada7-b0724d7bbc90](https://user-images.githubusercontent.com/37235023/222046336-4f35b4fb-5d3f-4971-a1d1-9dd990602a6c.png)

And this is the synthesized scribble map after preprocessor (you can use our scribble code to get this).
![221408976-40600115-72cf-4dad-afd1-f7dcbea5985d](https://user-images.githubusercontent.com/37235023/222046368-006049f0-e110-4c0c-bedd-0d91b20f60b7.png)

Then let me show off a bit my **prompt engineering skills**. I want a house under the winter snow. I will use this prompt:

 - Prompt:
```
Professional high-quality wide-angle digital art of a house designed by frank lloyd wright.
A delightful winter scene. photorealistic, epic fantasy, dramatic lighting, cinematic,
extremely high detail, cinematic lighting, trending on artstation, cgsociety, realistic
rendering of Unreal Engine 5, 8k, 4k, HQ, wallpaper
```
 - Negative Prompt:
```
longbody, lowres, bad anatomy, bad hands, missing fingers, extra digit, fewer digits,
cropped, worst quality, low quality
```

The **ControlNet-Self** is just our final released ControlNet and you can actually reproduce the results with below parameters. Note that we will just use same random seed `123456` for all experiments and generate 16 images without cherry-picking.

![221408989-4b00b150-a709-473f-85b8-02b6e2920fda](https://user-images.githubusercontent.com/37235023/222046887-d6fb4c37-c0e6-499b-ab7e-8e7ae071e84c.png)

### ControlNet-Self Results
![221408996-e9de3cdc-a8bb-4baf-9f59-bdf5566cfe0c](https://user-images.githubusercontent.com/37235023/222049182-ea320e76-5cb3-4836-bfe0-029e60acf1d3.png)

### ControlNet-Lite Results
![221409014-46fb7bdb-2c1f-4212-b97b-0134b1666301](https://user-images.githubusercontent.com/37235023/222049092-460ddc5c-7ade-4bb6-bfb0-1c8f674898e9.png)

### ControlNet-MLP Results
![221409020-20943cfe-12c3-4981-8a5b-1b38b21bdc1c](https://user-images.githubusercontent.com/37235023/222049171-5265d41b-b677-4b8c-bc34-71c628739cb0.png)

**Comment by NorbertZheng** [comment]
## Surprise, Surprise
It seems that they all give good results! The only difference is in some aesthetic concepts.

But why? Is the problem of controlling Stable Diffusion so trivial, and everything can work very well?

Why not turn off the ControlNet and see what happens:

![221409036-f8c85fb6-0162-47af-b0dd-b91d783d2918](https://user-images.githubusercontent.com/37235023/222050083-8c8a638f-8db7-40b8-8c54-0cfc9383f9f2.png)

Ah, then the secret trick is clear!

Because my prompts are carefully prepared, even without any control, the standard Stable Diffusion can already generate similar images that have many "overlapping concepts/semantics/shapes" with the input scribble maps.

In this case, it is true that every method can work very well.

In fact, in such an "easy" experimental setting, I believe [Sketch-Guided Diffusion](https://sketch-guided-diffusion.github.io/) or even [anisotropic filtering](https://en.wikipedia.org/wiki/Anisotropic_filtering) will also work very well to change the shape of objects and fit it to some user-specified structure.

But what about some other cases?

**Comment by NorbertZheng** [comment]
## The Non-Prompt Test
Here we must introduce the Non-Prompt Test (NPT), a test that can avoid the influence of the prompts and test the "pure" capability of ControlNet encoder.

NPT is simple - **just remove all prompts** (and put image conditions on the "c" side of cfg formulation "prd=uc+(c-uc)*cfg_scale" so that the cfg scale can still work). In our user interface, we call this ["Guess Mode"](https://github.com/lllyasviel/ControlNet#guess-anchor) because the model seems to guess contents from input control maps.

Because no prompt is available, the ControlNet encoder must recognize everything on its own. This is really challenging, and note that all our production-ready ControlNets have passed extensive NPT tests before we made them publicly available.

The "ControlNet-Self" is just our final released ControlNet and you can actually reproduce the results with below parameters. Note that we do not input any prompts.

![221409053-ad857843-3456-41b4-81d7-7f65da321bdd](https://user-images.githubusercontent.com/37235023/222050654-7fff476e-96ae-4ee1-8a71-adf56f52bddb.png)

### ControlNet-Self Results
![221409050-8d904d26-afe3-427b-bd41-7023dd9e6156](https://user-images.githubusercontent.com/37235023/222050771-0e6b62aa-03ba-45e1-a58c-12cbb647e90b.png)

### ControlNet-Lite Results
![221409061-6ac516f6-658f-49d2-8094-ddc6cae19836](https://user-images.githubusercontent.com/37235023/222050790-fb6bfe1a-4028-4d23-8b16-91bbe56ba79d.png)

### ControlNet-MLP Results
![221409064-e76ded89-1c6a-4c44-bdf3-da86050404bc](https://user-images.githubusercontent.com/37235023/222050815-54cc4b33-b38d-40a1-8253-4c6582c4a316.png)

**Comment by NorbertZheng** [comment]
## Observations
Now things are much clearer.
 - The difference between different encoders is in their capability to recognize contents in input control maps.
   - ControlNet-Self has strong recognition capability so that it works well even without prompts.
   - ControlNet-Lite and ControlNet-MLP are weak in this capability, and they cannot control Stable Diffusion (SD) to generate meaningful images without the help of user prompts.

### But, is this really important?
The answer depends on your goal.
 - If your goal is to build a method as robust as the production-ready ControlNets, this capability is important. In a production environment, we never know how strange user prompts will be, and user prompts are not likely to always cover everything in the control maps. We always want the encoder to have some recognition capability.
 - If your goal is to solve some specific problems in a research project, or if you have very aligned or fixed inputs, then perhaps you may consider some lightweight solutions (although I personally think the design of ControlNet-Self can also work well in this case).

But if you want to achieve a system with the quality similar to [Style2Paints V5](https://github.com/lllyasviel/style2paints/tree/master/V5_preview), then to the best of my knowledge, the ControlNet-Self is the only solution.

**Comment by NorbertZheng** [comment]
## Before We End
### Why we need these zero convolutions?
Now we also know why we need these zero convolutions
![221409333-652b2dcd-6074-4042-be7b-0bd08ba7cc42](https://user-images.githubusercontent.com/37235023/222051341-daac8fc3-32c2-418a-bdba-4bc7c7f7197d.png)


Just imagine that
 - these layers are initialized with noise, then a few training steps will immediately destroy the trainable copy.

The risk is very high that you are just training the already destroyed trainable copy from scratch again. To obtain the aforementioned object recognition capability would require extensive retraining -- similar to the amount of training required to produce the Stable Diffusion model itself.

### Whyy it is important that ControlNet encoder should receive prompts?
We also know why it is important that ControlNet encoder should also receive prompts:
![221409341-2ec6645d-45e3-414c-818b-d0a7b0f8324c](https://user-images.githubusercontent.com/37235023/222052306-764decc9-1130-4ec6-b38a-ff45d57915d7.png)

With this part, the ControlNet encoder's object recognition can be guided by the prompts so that
 - even when the prompts and the recognized control map semantics conflict, the user's prompt remains dominant.

For example, we already know that without prompts, the model can recognize the house in the house scribble map, but we can still make it into cakes: "delicious cakes" using that house scribble map
![221409080-01d93680-a61d-46c2-82ee-4a739152caa4](https://user-images.githubusercontent.com/37235023/222052544-274fb846-b155-4dca-ab0f-dd3c091aeb95.png)

Finally, note that this field is moving very fast and we won’t be surprised if some method suddenly comes out with just a few parameters and can also recognize objects equally well.
